lib/orderbook/OrderBook.ts

```diff
@@ -1,151 +1,220 @@
-      }
-      await Promise.all(updatedOrders);
+        this.updateOrderQuantity(this.ownOrders, maker, maker.quantity);
+      });
     }
     if (remainingOrder) {
       this.broadcastOrder(remainingOrder);
 
       if (this.config.internalmatching) {
-        const dbOrder = await this.repository.addOrder(<db.OrderFactory>remainingOrder);
-        this.logger.debug(`order added: ${JSON.stringify(dbOrder)}`);
+        this.addOrder(this.ownOrders, remainingOrder);
+        this.logger.debug(`order added: ${JSON.stringify(remainingOrder)}`);
       }
     }
 
     return matchingResult;
   }
 
-  private addPeerOrder = async (order: orders.PeerOrder): Promise<void> => {
+  private addPeerOrder = (order: orders.PeerOrder) => {
     const matchingEngine = this.matchingEngines[order.pairId];
     if (!matchingEngine) {
       this.logger.debug(`incoming peer order invalid pairId: ${order.pairId}`);
       return;
     }
 
-    const stampedOrder: orders.StampedPeerOrder = { ...order, createdAt: new Date() };
+    const stampedOrder: orders.StampedPeerOrder = { ...order, createdAt: ms() };
     matchingEngine.addPeerOrder(stampedOrder);
-    const dbOrder = await this.repository.addOrder(stampedOrder);
-    this.logger.debug(`order added: ${JSON.stringify(dbOrder)}`);
+    this.addOrder(this.peerOrders, stampedOrder);
+    this.logger.debug(`order added: ${JSON.stringify(stampedOrder)}`);
+  }
+
+  private updateOrderQuantity = (type: { [pairId: string]: Orders }, order: orders.StampedOrder, decreasedQuantity: number) => {
+    const object = this.getOrderMap(type, order);
+    object[order.id].quantity = object[order.id].quantity - decreasedQuantity;
+    if (object[order.id].quantity === 0) {
+      delete object[order.id];
+    }
+  }
+
+  private addOrder = (type: { [pairId: string]: Orders }, order: orders.StampedOrder) => {
+    this.getOrderMap(type, order)[order.id] = order;
+  }
+
+  private getOrderMap = (type: { [pairId: string]: Orders }, order: orders.StampedOrder): { [ id: string ]: orders.StampedOrder } => {
+    const orders = type[order.pairId];
+    if (order.quantity > 0) {
+      return orders.buyOrders;
+    } else {
+      return orders.sellOrders;
+    }
   }
 
   private broadcastOrder = async (order: orders.StampedOwnOrder): Promise<void> => {
-    if (!this.pool) {
-      return;
+    if (this.pool) {
+      const outgoingOrder = await this.castToOutgoingOrder(order);
+
+      if (outgoingOrder) {
+        this.pool.broadcastOrder(outgoingOrder);
+      }
     }
+  }
+
+  private castToOutgoingOrder = async (order: orders.StampedOwnOrder): Promise<orders.OutgoingOrder | void> => {
     const invoice = await this.getInvoice(order);
-    if (!invoice) {
-      return;
-    }
 
-    const { createdAt, ...outgoingOrder } = { ...order, invoice };
-    this.pool.broadcastOrder(outgoingOrder);
+    if (!invoice) return;
+
+    const { ...outgoingOrder } = { ...order, invoice };
```

Comment:
This can be written as `const outgoingOrder = { ...order, invoice }`;

But any reason for not dropping the `createdAt`, as it previously was? (since it's not defined in `OutgoingOrder` type)

Reply:
I though that it gets dropped automatically because it is not defined in the type?

Reply:
It doesn't do it, so you can easily end-up with having a run-time object which has more properties then what was defined in TS.